src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,248 @@
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    /// <summary>
+    /// CA2016: Forward CancellationToken to async methods.
+    /// 
+    /// Conditions for positive cases:
+    ///     - The containing method signature receives a ct parameter. It can be a method, a nested method, an action or a func.
+    ///     - The invocation is not receiving a ct argument, and...
+    ///     - The invocation method either:
+    ///         - Has no overloads but its current signature receives an optional ct=default, being used right now, or...
+    ///         - Has a method overload with the exact same arguments in the same order, plus one ct parameter at the end.
+    ///         
+    /// Conditions for negative cases:
+    ///     - The containing method signature does not receive a ct parameter.
+    ///     - The invocation is explicitly receiving a ct argument.
+    ///     - The invocation method signature receives a ct but one is already being explicitly passed, or...
+    ///     - The invocation method does not have an overload with the exact same arguments that also receives a ct, or...
+    ///     - The invocation method has overloads that receive more than one ct.
+    ///
+    /// Future improvements:
+    ///     - Finding an overload with one ct parameter, but not in last position.
+    ///     - Passing a named ct in a different position than default.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class ForwardCancellationTokenToAsyncMethodsAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2016";
+
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToAsyncMethodsDescription),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToAsyncMethodsMessage),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToAsyncMethodsTitle),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+        internal static DiagnosticDescriptor ForwardCancellationTokenToAsyncMethodsRule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
```

Comment:
Can you confirm this has been approved as a build warning?

Reply:
I forwarded this question to my team. Waiting for a confirmation.

Reply:
Have we run this yet over various codebases like dotnet/runtime and dotnet/aspnetcore to see what kind of false positives it might have?  We should probably start with it as informational until we're sure the false positive rate is basically zero.

Reply:
Okay sure, I'll do that then.

Reply:
Ok so below is a screenshot of the diagnostic in the runtime repo (I purposedly removed the passed `cancellationToken` argument). Notice that the method name is what is being underlined, not the whole invocation 😄. I clicked it and it was properly fixed.

I'm building all runtime to see what cases we find.

![image](https://user-images.githubusercontent.com/1175054/83714916-00ed5580-a5e1-11ea-8f72-b11500e977db.png)

Reply:
This warning was hit 63 times in the dotnet/runtime repo. I'll prepare a PR with the applied fixes so we can review them.

Reply:
Found a bug in this analyzer.

The file `runtime\src\libraries\Microsoft.Extensions.Hosting.Abstractions\src\BackgroundService.cs` is getting an incorrect flag when calling `CancellationTokenSource.CreateLinkedTokenSource`:

![image](https://user-images.githubusercontent.com/1175054/83720261-942c8800-a5ed-11ea-8baa-062ea249bdf8.png)

The method has two unusual overloads:

```cs
namespace System.Threading
{
    public class CancellationTokenSource : IDisposable
    {
        public static CancellationTokenSource CreateLinkedTokenSource(CancellationToken token);
        public static CancellationTokenSource CreateLinkedTokenSource(CancellationToken token1, CancellationToken token2);
        public static CancellationTokenSource CreateLinkedTokenSource(params CancellationToken[] tokens);
    }
}
```

I will investigate tomorrow why this is getting a false positive.

Reply:
Found another one, kinda puzzling:
![image](https://user-images.githubusercontent.com/1175054/83721330-b58e7380-a5ef-11ea-8317-df4872814665.png)

That `ReadAsStreamAsync` method does not have an overload that takes a `CancellationToken`. In fact, it has no overloads.

Reply:
Found another interesting case, not too bad but I want to fix it - In some cases, the fixer adds the cancellationToken argument with a name, but I only wanted to do that when other arguments were already named.

Example: `runtime\src\libraries\Common\src\System\Net\Http\aspnetcore\Quic\Implementations\Mock\MockListener.cs`

![image](https://user-images.githubusercontent.com/1175054/83787816-329a0700-a649-11ea-8e16-fdec190980ab.png)

Reply:
Another case: This lambda is not getting passed a token as an argument. Instead, the token is created inside the lambda. It should not have been diagnosed by this analyzer:

`\runtime\src\libraries\System.Net.Http\src\System\Net\Http\MessageProcessingHandler.cs`"
![image](https://user-images.githubusercontent.com/1175054/83788780-9670ff80-a64a-11ea-8aa7-5944dc5ec3d7.png)

Reply:
Found another fixer bug.
When calling a method using nullability for the object, the substitution is not done correctly - the method name gets unexpectedly substituted with the object name:

File: `runtime\src\libraries\System.Threading.Channels\src\System\Threading\Channels\SingleConsumerUnboundedChannel.cs`

Before:
![image](https://user-images.githubusercontent.com/1175054/83790373-f23c8800-a64c-11ea-98d7-bf92e7ab508c.png)

After:
![image](https://user-images.githubusercontent.com/1175054/83790428-08e2df00-a64d-11ea-889d-e58eabba4f5c.png)

Reply:
Found a case that the fixer can't handle - If the invocation is inside a `lock` block, it doesn't know where to get the token argument from, so the fixer gives no options:

`runtime\src\libraries\System.Threading.Tasks.Dataflow\src\Base\DataflowBlock.cs`

![image](https://user-images.githubusercontent.com/1175054/83790937-cf5ea380-a64d-11ea-851c-aade7cbaaef1.png)

Reply:
> That ReadAsStreamAsync method does not have an overload that takes a CancellationToken. In fact, it has no overloads.

It does in .NET 5.  Maybe the analyzer is firing for a build of that consuming project targeting net5.0?

> Found a case that the fixer can't handle

It's ok for the analyzer to flag cases the fixer can't fix as long as a) the case is valid and b) we don't offer the fixer.

Reply:
>> _That ReadAsStreamAsync method does not have an overload that takes a CancellationToken. In fact, it has no overloads._
> _It does in .NET 5. Maybe the analyzer is firing for a build of that consuming project targeting net5.0?_
The instructions I followed to compile the roslyn-analyzers project used this environment variable: `RUNTIMEPACKAGEVERSION=3.0.0`. Do you think that could be the reason why the fixer was unable to find an overload that takes a token?

> _It's ok for the analyzer to flag cases the fixer can't fix as long as a) the case is valid and b) we don't offer the fixer._

I suspect it wouldn't be too hard to get this case handled, let me try to fix it.

Reply:
More detailed explanation in the dotnet/runtime PR: https://github.com/dotnet/runtime/pull/37607